### DISCUSSION TITLE
How to do CRTP in CPP2?

### DISCUSSION THREAD

**Discussion by jamadagni** [hsutter/cppfront#1243] [Q&A]
Hello. I have placed a minimal example of CRTP [here](https://gist.github.com/jamadagni/f82da0d618d36ca74788d3472dc8f8c2). I would like to know whether CPP2's Awesome Metafunction Powers ™ :slightly_smiling_face:  (or any other powers) would enable it to do CRTP like this.

I didn't find any mention of CRTP by searching this repo. (Hope I didn't miss anything.)

Expected output:

    मेषः सिंहः
    5
    अश्वयुजौ पू॰ फ॰ रोहिणी रेवती

If there are any difficulties with the Devanagari text (hope not) the Latin abbreviated alternates can be used instead and the output would be:

    Mes Sim
    5
    Asvn PPha Rohi Reva

I am at the verge of seriously starting on a project, and would like to do it using CPP2 if possible, but CRTP is important for me…

**Comment by bluetarpmedia** [comment]
Here's an example of CRTP in Cpp2:
```cpp
printer: <T> type = {
    print: (this) = {
        derived:= unsafe_cast<*const T>(this&);
        name:=    derived*.get_name();
        std::cout << "The animal is a (name)$\n";
    }
}

cat: type = {
    this: printer<cat>;
    get_name: (this) -> std::string = { return "cat"; }
}

dog: type = {
    this: printer<dog>;
    get_name: (this) -> std::string = { return "dog"; }
}

main: () -> int = {
    c: cat = ();
    d: dog = ();
    c.print();
    d.print();
    return 0;
}
```

[Run on Compiler Explorer](https://cpp2.godbolt.org/z/d7MYse454)

There's also an [open suggestion](https://github.com/hsutter/cppfront/issues/1197) to implement a "deducing this" equivalent in Cpp2, which would simplify CRTP.

@hsutter I found that this syntax `unsafe_cast<*const T>(this&)` was the only way I could cast from the base down to the derived type. Is that correct or is there a better way?

**Reply by jamadagni** [reply]
Hmm if there is no requirement to access the actual derived object I guess the following would do (and works, thanks):

```
printer: <T> type = {
    print: (this) = {
        std::cout << "The animal is a (T::get_name())$\n";
    }
}

cat: type = {
    this: printer<cat>;
    get_name: () "cat";
}

dog: type = {
    this: printer<dog>;
    get_name: () "dog";
}

main: () = {
    c: cat = (); d: dog = ();
    c.print(); d.print();
}
```

**Comment by MaxSagebaum** [comment]
> I am at the verge of seriously starting on a project, and would like to do it using CPP2 if possible, but CRTP is important for me…

It is nice that you want to use cppfront for this. If this is more than an exploration project you should not use it. Cppfront is still in development and there are a lot of corner cases that do not work yet.
